File: go/http_client.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// TestHTTP 使用原生 HTTP 客户端测试（类似 Python requests / Java OkHttp）
type TestHTTP struct {
	config *Config
	client *http.Client
}

// NewTestHTTP 创建 HTTP 测试实例
func NewTestHTTP(config *Config) *TestHTTP {
	return &TestHTTP{
		config: config,
		client: &http.Client{
			Timeout: 60 * time.Second,
		},
	}
}

type chatRequest struct {
	Model    string    `json:"model"`
	Messages []message `json:"messages"`
	MaxTokens int      `json:"max_tokens"`
	Stream   bool      `json:"stream"`
}

type message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type chatResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
}

type chatChunk struct {
	Choices []struct {
		Delta struct {
			Content string `json:"content"`
		} `json:"delta"`
	} `json:"choices"`
}

type modelsResponse struct {
	Data []struct {
		ID string `json:"id"`
	} `json:"data"`
}

// TestNonStream 测试非流式 Chat Completions
func (t *TestHTTP) TestNonStream() bool {
	fmt.Println("\n--- Chat Completions (非流式) ---")

	payload := chatRequest{
		Model: t.config.Model,
		Messages: []message{
			{Role: "user", Content: "你好，请回复'连通成功'"},
		},
		MaxTokens: 100,
		Stream:    false,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		fmt.Printf("❌ JSON 编码失败: %v\n", err)
		return false
	}

	req, err := http.NewRequest("POST", t.config.BaseURL+"/chat/completions", strings.NewReader(string(body)))
	if err != nil {
		fmt.Printf("❌ 请求创建失败: %v\n", err)
		return false
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+t.config.APIKey)

	start := time.Now()
	resp, err := t.client.Do(req)
	if err != nil {
		fmt.Printf("❌ 请求失败: %v\n", err)
		return false
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		fmt.Printf("❌ HTTP %d: %s\n", resp.StatusCode, string(respBody))
		return false
	}

	var result chatResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		fmt.Printf("❌ JSON 解码失败: %v\n", err)
		return false
	}

	elapsed := time.Since(start)
	if len(result.Choices) > 0 {
		fmt.Printf("回复: %s\n", result.Choices[0].Message.Content)
		fmt.Printf("✅ 成功 | 耗时: %.3fs\n", elapsed.Seconds())
		return true
	}

	fmt.Println("❌ 无响应内容")
	return false
}

// TestStream 测试流式 Chat Completions
func (t *TestHTTP) TestStream() bool {
	fmt.Println("\n--- Chat Completions (流式) ---")

	payload := chatRequest{
		Model: t.config.Model,
		Messages: []message{
			{Role: "user", Content: "你好，请回复'连通成功'"},
		},
		MaxTokens: 100,
		Stream:    true,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		fmt.Printf("❌ JSON 编码失败: %v\n", err)
		return false
	}

	req, err := http.NewRequest("POST", t.config.BaseURL+"/chat/completions", strings.NewReader(string(body)))
	if err != nil {
		fmt.Printf("❌ 请求创建失败: %v\n", err)
		return false
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+t.config.APIKey)

	start := time.Now()
	resp, err := t.client.Do(req)
	if err != nil {
		fmt.Printf("❌ 请求失败: %v\n", err)
		return false
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		fmt.Printf("❌ HTTP %d: %s\n", resp.StatusCode, string(respBody))
		return false
	}

	var firstTokenTime *float64
	var fullText strings.Builder
	scanner := bufio.NewScanner(resp.Body)

	for scanner.Scan() {
		line := scanner.Text()
		if !strings.HasPrefix(line, "data: ") {
			continue
		}
		data := strings.TrimPrefix(line, "data: ")
		if data == "[DONE]" {
			break
		}

		var chunk chatChunk
		if err := json.Unmarshal([]byte(data), &chunk); err != nil {
			continue
		}

		if len(chunk.Choices) > 0 {
			content := chunk.Choices[0].Delta.Content
			if content != "" {
				if firstTokenTime == nil {
					ttft := time.Since(start).Seconds()
					firstTokenTime = &ttft
				}
				fullText.WriteString(content)
				fmt.Print(content)
			}
		}
	}

	if fullText.Len() > 0 {
		elapsed := time.Since(start).Seconds()
		ttft := 0.0
		if firstTokenTime != nil {
			ttft = *firstTokenTime
		}
		fmt.Printf("\n✅ 成功 | TTFT: %.3fs | 总耗时: %.3fs\n", ttft, elapsed)
		return true
	}

	fmt.Println("\n❌ 无响应内容")
	return false
}

// TestModelsList 测试 Models 列表
func (t *TestHTTP) TestModelsList() bool {
	fmt.Println("\n--- Models List ---")

	req, err := http.NewRequest("GET", t.config.BaseURL+"/models", nil)
	if err != nil {
		fmt.Printf("❌ 请求创建失败: %v\n", err)
		return false
	}
	req.Header.Set("Authorization", "Bearer "+t.config.APIKey)

	resp, err := t.client.Do(req)
	if err != nil {
		fmt.Printf("❌ 请求失败: %v\n", err)
		return false
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		fmt.Printf("❌ HTTP %d: %s\n", resp.StatusCode, string(respBody))
		return false
	}

	var result modelsResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		fmt.Printf("❌ JSON 解码失败: %v\n", err)
		return false
	}

	fmt.Printf("可用模型 (%d 个):\n", len(result.Data))
	for _, m := range result.Data {
		fmt.Printf("  - %s\n", m.ID)
	}
	fmt.Println("✅ 成功")
	return true
}

File: rust/src/http_client.rs
use crate::config::Config;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::time::Instant;

type BoxError = Box<dyn std::error::Error + Send + Sync>;

#[derive(Serialize)]
struct ChatRequest {
    model: String,
    messages: Vec<Message>,
    max_tokens: u32,
    #[serde(skip_serializing_if = "std::ops::Not::not")]
    stream: bool,
}

#[derive(Serialize)]
struct Message {
    role: String,
    content: String,
}

#[derive(Deserialize)]
struct ChatResponse {
    choices: Vec<Choice>,
}

#[derive(Deserialize)]
struct Choice {
    message: ResponseMessage,
}

#[derive(Deserialize)]
struct ResponseMessage {
    content: String,
}

#[derive(Deserialize)]
struct ChatChunk {
    choices: Vec<ChunkChoice>,
}

#[derive(Deserialize)]
struct ChunkChoice {
    delta: Delta,
}

#[derive(Deserialize)]
struct Delta {
    content: Option<String>,
}

#[derive(Deserialize)]
struct ModelsResponse {
    data: Vec<ModelInfo>,
}

#[derive(Deserialize)]
struct ModelInfo {
    id: String,
}

pub struct HttpClient {
    config: Config,
    client: Client,
}

impl HttpClient {
    pub fn new(config: Config) -> Self {
        let client = Client::builder()
            .timeout(std::time::Duration::from_secs(60))
            .build()
            .expect("Failed to build HTTP client");

        HttpClient { config, client }
    }

    fn build_chat_payload(&self, content: &str, stream: bool) -> ChatRequest {
        ChatRequest {
            model: self.config.model.clone(),
            messages: vec![Message {
                role: "user".to_string(),
                content: content.to_string(),
            }],
            max_tokens: 100,
            stream,
        }
    }

    pub async fn test_models_list(&self) -> bool {
        println!("\n--- Models List ---");

        let result: Result<(), BoxError> = async {
            let resp = self
                .client
                .get(format!("{}/models", self.config.base_url))
                .bearer_auth(&self.config.api_key)
                .send()
                .await?;

            if !resp.status().is_success() {
                let status = resp.status();
                let body = resp.text().await?;
                return Err(format!("HTTP {}: {}", status, body).into());
            }

            let models: ModelsResponse = resp.json().await?;
            println!("可用模型 ({} 个):", models.data.len());
            for m in &models.data {
                println!("  - {}", m.id);
            }
            println!("✅ 成功");
            Ok(())
        }
        .await;

        match result {
            Ok(()) => true,
            Err(e) => {
                println!("❌ {}", e);
                false
            }
        }
    }

    pub async fn test_non_stream(&self) -> bool {
        println!("\n--- Chat Completions (非流式) ---");

        let result: Result<(), BoxError> = async {
            let body = self.build_chat_payload("你好，请回复'连通成功'", false);
            let start = Instant::now();

            let resp = self
                .client
                .post(format!("{}/chat/completions", self.config.base_url))
                .bearer_auth(&self.config.api_key)
                .json(&body)
                .send()
                .await?;

            if !resp.status().is_success() {
                let status = resp.status();
                let body_text = resp.text().await?;
                return Err(format!("HTTP {}: {}", status, body_text).into());
            }

            let data: ChatResponse = resp.json().await?;
            let elapsed = start.elapsed();

            if let Some(choice) = data.choices.first() {
                println!("回复: {}", choice.message.content);
                println!("✅ 成功 | 耗时: {:.3}s", elapsed.as_secs_f64());
                Ok(())
            } else {
                Err("无响应内容".into())
            }
        }
        .await;

        match result {
            Ok(()) => true,
            Err(e) => {
                println!("❌ {}", e);
                false
            }
        }
    }

    pub async fn test_stream(&self) -> bool {
        println!("\n--- Chat Completions (流式) ---");

        let result: Result<(), BoxError> = async {
            let body = self.build_chat_payload("你好，请回复'连通成功'", true);
            let start = Instant::now();

            let resp = self
                .client
                .post(format!("{}/chat/completions", self.config.base_url))
                .bearer_auth(&self.config.api_key)
                .json(&body)
                .send()
                .await?;

            if !resp.status().is_success() {
                let status = resp.status();
                let body_text = resp.text().await?;
                return Err(format!("HTTP {}: {}", status, body_text).into());
            }

            let mut first_token_time: Option<f64> = None;
            let mut full_text = String::new();
            let mut stream = resp.bytes_stream();

            use futures::StreamExt;
            let mut buf = String::new();

            while let Some(chunk) = stream.next().await {
                let bytes = chunk?;
                buf.push_str(&String::from_utf8_lossy(&bytes));

                // 按行处理 SSE
                while let Some(newline_pos) = buf.find('\n') {
                    let line = buf[..newline_pos].trim().to_string();
                    buf = buf[newline_pos + 1..].to_string();

                    if !line.starts_with("data: ") {
                        continue;
                    }
                    let data = &line[6..];
                    if data == "[DONE]" {
                        break;
                    }

                    if let Ok(chunk) = serde_json::from_str::<ChatChunk>(data) {
                        if let Some(choice) = chunk.choices.first() {
                            if let Some(content) = &choice.delta.content {
                                if !content.is_empty() {
                                    if first_token_time.is_none() {
                                        first_token_time = Some(start.elapsed().as_secs_f64());
                                    }
                                    full_text.push_str(content);
                                    print!("{}", content);
                                    use std::io::Write;
                                    std::io::stdout().flush().ok();
                                }
                            }
                        }
                    }
                }
            }

            if !full_text.is_empty() {
                let elapsed = start.elapsed().as_secs_f64();
                let ttft = first_token_time.unwrap_or(0.0);
                println!("\n✅ 成功 | TTFT: {:.3}s | 总耗时: {:.3}s", ttft, elapsed);
                Ok(())
            } else {
                Err("无响应内容".into())
            }
        }
        .await;

        match result {
            Ok(()) => true,
            Err(e) => {
                println!("❌ {}", e);
                false
            }
        }
    }
}
